man/ibd2sdi.1
'\" t
.\"     Title: ibd2sdi
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/25/2026
.\"    Manual: MySQL Database System
.\"    Source: MySQL 9.7
.\"  Language: English
.\"
.TH "IBD2SDI" "1" "06/25/2026" "MySQL 9\&.7" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ibd2sdi \- InnoDB utility for extracting serialized dictionary information (SDI) from an InnoDB tablespace
.SH "SYNOPSIS"
.HP \w'\fBibd2sdi\ [\fR\fB\fIoptions\fR\fR\fB]\ \fR\fB\fIfile_name1\fR\fR\fB\ [\fR\fB\fIfile_name2\ file_name3\ \&.\&.\&.]\fR\fR\ 'u
\fBibd2sdi [\fR\fB\fIoptions\fR\fR\fB] \fR\fB\fIfile_name1\fR\fR\fB [\fR\fB\fIfile_name2 file_name3 \&.\&.\&.]\fR\fR
.SH "DESCRIPTION"
.PP
\fBibd2sdi\fR
is a utility for extracting
\m[blue]\fBserialized dictionary information\fR\m[]\&\s-2\u[1]\d\s+2
(SDI) from
InnoDB
tablespace files\&. SDI data is present in all persistent
InnoDB
tablespace files\&.
.PP
\fBibd2sdi\fR
can be run on
file\-per\-table
tablespace files (*\&.ibd
files),
general tablespace
files (*\&.ibd
files),
\m[blue]\fBsystem tablespace\fR\m[]\&\s-2\u[2]\d\s+2
files (ibdata*
files), and the data dictionary tablespace (mysql\&.ibd)\&. It is not supported for use with temporary tablespaces or undo tablespaces\&.
.PP
\fBibd2sdi\fR
can be used at runtime or while the server is offline\&. During
DDL
operations,
ROLLBACK
operations, and undo log purge operations related to SDI, there may be a short interval of time when
\fBibd2sdi\fR
fails to read SDI data stored in the tablespace\&.
.PP
\fBibd2sdi\fR
performs an uncommitted read of SDI from the specified tablespace\&. Redo logs and undo logs are not accessed\&.
.PP
Invoke the
\fBibd2sdi\fR
utility like this:
.sp
.if n \{\
.RS 4
.\}
.nf
ibd2sdi [\fIoptions\fR] \fIfile_name1\fR [\fIfile_name2 file_name3 \&.\&.\&.\fR]
.fi
.if n \{\
.RE
.\}
.PP
\fBibd2sdi\fR
supports multi\-file tablespaces like the
InnoDB
system tablespace, but it cannot be run on more than one tablespace at a time\&. For multi\-file tablespaces, specify each file:
.sp
.if n \{\
.RS 4
.\}
.nf
ibd2sdi ibdata1 ibdata2
.fi
.if n \{\
.RE
.\}
.PP
The files of a multi\-file tablespace must be specified in order of the ascending page number\&. If two successive files have the same space ID, the later file must start with the last page number of the previous file + 1\&.
.PP
\fBibd2sdi\fR
outputs SDI (containing id, type, and data fields) in
JSON
format\&.
ibd2sdi Options
.PP
\fBibd2sdi\fR
supports the following options:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-help\fR,
\fB\-h\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--help
T}
T{
Type
T}:T{
Boolean
T}
T{
Default Value
T}:T{
false
T}
.TE
.sp 1
Display a help message and exit\&. For example:
.sp
.if n \{\
.RS 4
.\}
.nf
Usage: \&./ibd2sdi [\-v] [\-c <strict\-check>] [\-d <dump file name>] [\-n] filename1 [filenames]
See http://dev\&.mysql\&.com/doc/refman/9\&.7/en/ibd2sdi\&.html for usage hints\&.
  \-h, \-\-help          Display this help and exit\&.
  \-v, \-\-version       Display version information and exit\&.
  \-#, \-\-debug[=name]  Output debug log\&. See
                      http://dev\&.mysql\&.com/doc/refman/9\&.7/en/dbug\-package\&.html
  \-d, \-\-dump\-file=name
                      Dump the tablespace SDI into the file passed by user\&.
                      Without the filename, it will default to stdout
  \-s, \-\-skip\-data     Skip retrieving data from SDI records\&. Retrieve only id
                      and type\&.
  \-i, \-\-id=#          Retrieve the SDI record matching the id passed by user\&.
  \-t, \-\-type=#        Retrieve the SDI records matching the type passed by
                      user\&.
  \-c, \-\-strict\-check=name
                      Specify the strict checksum algorithm by the user\&.
                      Allowed values are innodb, crc32, none\&.
  \-n, \-\-no\-check      Ignore the checksum verification\&.
  \-p, \-\-pretty        Pretty format the SDI output\&.If false, SDI would be not
                      human readable but it will be of less size
                      (Defaults to on; use \-\-skip\-pretty to disable\&.)
Variables (\-\-variable\-name=value)
and boolean options {FALSE|TRUE}  Value (after reading options)
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\- \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
debug                             (No default value)
dump\-file                         (No default value)
skip\-data                         FALSE
id                                0
type                              0
strict\-check                      crc32
no\-check                          FALSE
pretty                            TRUE
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-version\fR,
\fB\-v\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--version
T}
T{
Type
T}:T{
Boolean
T}
T{
Default Value
T}:T{
false
T}
.TE
.sp 1
Display version information and exit\&. For example:
.sp
.if n \{\
.RS 4
.\}
.nf
ibd2sdi  Ver 9\&.7\&.0 for Linux on x86_64 (Source distribution)
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-debug[=\fR\fB\fIdebug_options\fR\fR\fB]\fR,
\fB\-# [\fR\fB\fIdebug_options\fR\fR\fB]\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--debug=options
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Prints a debug log\&. For debug options, refer to
Section\ \&7.9.4, \(lqThe DBUG Package\(rq\&.
.sp
.if n \{\
.RS 4
.\}
.nf
ibd2sdi \-\-debug=d:t /tmp/ibd2sdi\&.trace
.fi
.if n \{\
.RE
.\}
.sp
This option is available only if MySQL was built using
\fBWITH_DEBUG\fR\&. MySQL release binaries provided by Oracle are
\fInot\fR
built using this option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-dump\-file=\fR,
\fB\-d\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--dump-file=file
T}
T{
Type
T}:T{
File name
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Dumps serialized dictionary information (SDI) into the specified dump file\&. If a dump file is not specified, the tablespace SDI is dumped to
stdout\&.
.sp
.if n \{\
.RS 4
.\}
.nf
ibd2sdi \-\-dump\-file=\fIfile_name\fR \&.\&./data/test/t1\&.ibd
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-skip\-data\fR,
\fB\-s\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--skip-data
T}
T{
Type
T}:T{
Boolean
T}
T{
Default Value
T}:T{
false
T}
.TE
.sp 1
Skips retrieval of
data
field values from the serialized dictionary information (SDI) and only retrieves the
id
and
type
field values, which are primary keys for SDI records\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBibd2sdi \-\-skip\-data \&.\&./data/test/t1\&.ibd\fR
["ibd2sdi"
,
{
	"type": 1,
	"id": 330
}
,
{
	"type": 2,
	"id": 7
}
]
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-id=\fR\fB\fI#\fR\fR,
\fB\-i \fR\fB\fI#\fR\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--id=#
T}
T{
Type
T}:T{
Integer
T}
T{
Default Value
T}:T{
0
T}
.TE
.sp 1
Retrieves serialized dictionary information (SDI) matching the specified table or tablespace object id\&. An object id is unique to the object type\&. Table and tablespace object IDs are also found in the
id
column of the
mysql\&.tables
and
mysql\&.tablespace
data dictionary tables\&. For information about data dictionary tables, see
Section\ \&16.1, \(lqData Dictionary Schema\(rq\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBibd2sdi \-\-id=7 \&.\&./data/test/t1\&.ibd\fR
["ibd2sdi"
,
{
	"type": 2,
	"id": 7,
	"object":
		{
    "mysqld_version_id": 80003,
    "dd_version": 80003,
    "sdi_version": 1,
    "dd_object_type": "Tablespace",
    "dd_object": {
        "name": "test/t1",
        "comment": "",
        "options": "",
        "se_private_data": "flags=16417;id=2;server_version=80003;space_version=1;",
        "engine": "InnoDB",
        "files": [
            {
                "ordinal_position": 1,
                "filename": "\&./test/t1\&.ibd",
                "se_private_data": "id=2;"
            }
        ]
    }
}
}
]
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-type=\fR\fB\fI#\fR\fR,
\fB\-t \fR\fB\fI#\fR\fR
.TS
allbox tab(:);
lB l
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--type=#
T}
T{
Type
T}:T{
Enumeration
T}
T{
Default Value
T}:T{
0
T}
T{
Valid Values
T}:T{
.PP
1
.PP
2
T}
.TE
.sp 1
Retrieves serialized dictionary information (SDI) matching the specified object type\&. SDI is provided for table (type=1) and tablespace (type=2) objects\&.
.sp
This example shows output for a tablespace
ts1
in the
test
database:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBibd2sdi \-\-type=2 \&.\&./data/test/ts1\&.ibd\fR
["ibd2sdi"
,
{
	"type": 2,
	"id": 7,
	"object":
		{
    "mysqld_version_id": 80003,
    "dd_version": 80003,
    "sdi_version": 1,
    "dd_object_type": "Tablespace",
    "dd_object": {
        "name": "test/ts1",
        "comment": "",
        "options": "",
        "se_private_data": "flags=16417;id=2;server_version=80003;space_version=1;",
        "engine": "InnoDB",
        "files": [
            {
                "ordinal_position": 1,
                "filename": "\&./test/ts1\&.ibd",
                "se_private_data": "id=2;"
            }
        ]
    }
}
}
]
.fi
.if n \{\
.RE
.\}
.sp
Due to the way in which
InnoDB
handles default value metadata, a default value may be present and non\-empty in
\fBibd2sdi\fR
output for a given table column even if it is not defined using
DEFAULT\&. Consider the two tables created using the following statements, in the database named
i:
.sp
.if n \{\
.RS 4
.\}
.nf
CREATE TABLE t1 (c VARCHAR(16) NOT NULL);
CREATE TABLE t2 (c VARCHAR(16) NOT NULL DEFAULT "Sakila");
.fi
.if n \{\
.RE
.\}
.sp
Using
\fBibd2sdi\fR, we can see that the
default_value
for column
c
is nonempty and is in fact padded to length in both tables, like this:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBibd2sdi \&.\&./data/i/t1\&.ibd  | grep \-m1 \*(Aq\e"default_value\e"\*(Aq | cut \-b34\- | sed \-e s/,//\fR
"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\enAAAAAAAAAAA="
$> \fBibd2sdi \&.\&./data/i/t2\&.ibd  | grep \-m1 \*(Aq\e"default_value\e"\*(Aq | cut \-b34\- | sed \-e s/,//\fR
"BlNha2lsYQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\enAAAAAAAAAAA="
.fi
.if n \{\
.RE
.\}
.sp
Examination of
\fBibd2sdi\fR
output may be easier using a JSON\-aware utility like
\fB\m[blue]\fBjq\fR\m[]\&\s-2\u[3]\d\s+2\fR, as shown here:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBibd2sdi \&.\&./data/i/t1\&.ibd  | jq \*(Aq\&.[1]["object"]["dd_object"]["columns"][0]["default_value"]\*(Aq\fR
"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\enAAAAAAAAAAA="
$> \fBibd2sdi \&.\&./data/i/t2\&.ibd  | jq \*(Aq\&.[1]["object"]["dd_object"]["columns"][0]["default_value"]\*(Aq\fR
"BlNha2lsYQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\enAAAAAAAAAAA="
.fi
.if n \{\
.RE
.\}
.sp
For more information, see the
\m[blue]\fBMySQL Internals documentation\fR\m[]\&\s-2\u[4]\d\s+2\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-strict\-check\fR,
\fB\-c\fR
.TS
allbox tab(:);
lB l
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--strict-check=algorithm
T}
T{
Type
T}:T{
Enumeration
T}
T{
Default Value
T}:T{
crc32
T}
T{
Valid Values
T}:T{
.PP
crc32
.PP
innodb
.PP
none
T}
.TE
.sp 1

ChangeLog
2010-03-18	The InnoDB Team

	* CMakeLists.txt:
	Fix Bug#52102 InnoDB Plugin shows performance drop compared to
	InnoDB (Windows)

2010-03-18	The InnoDB Team

	* buf0buf.ic:
	When comparing the time of the first access to a block against
	innodb_old_blocks_time, use 32-bit arithmetics. The comparison was
	incorrect on 64-bit systems.

2010-03-11	The InnoDB Team

	* buf0buf.h, buf0buf.ic:
	Fix and clarify the latching of some buf_block_t members.
	Note that check_index_page_at_flush is not protected by any mutex.
	Note and assert that lock_hash_val is protected by the rw-latch.

2010-03-10	The InnoDB Team

	* trx/trx0sys.c:
	Fix Bug#51653 outdated reference to set-variable

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb_bug21704.result,
	mysql-test/innodb_bug47621.result, mysql-test/innodb_bug47621.test:
	Fix Bug#47621 MySQL and InnoDB data dictionaries will become out of
	sync when renaming columns

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc:
	Fix Bug#51356 Many Valgrind errors in error messages
	with concurrent DDL

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, handler/handler0alter.cc,
	mysql-test/innodb_bug51378.result, mysql-test/innodb_bug51378.test:
	Fix Bug#51378 Init 'ref_length' to correct value, in case an out
	of bound MySQL primary_key

2010-03-10	The InnoDB Team

	* log/log0recv.c:
	Remove a bogus assertion about page numbers exceeding 0x90000000
	in the redo log. Abort when encountering a corrupted redo log
	record, unless innodb_force_recovery is set.

2010-03-09	The InnoDB Team

	* handler/ha_innodb.cc:
	Make SHOW ENGINE INNODB MUTEX STATUS display SUM(os_waits)
	for the buffer pool block mutexes and locks.

2010-03-08	The InnoDB Team

	* fil/fil0fil.c:
	Fix ALTER TABLE ... IMPORT TABLESPACE of compressed tables.

2010-03-03	The InnoDB Team

	* handler/handler0alter.cc, innodb-index.result, innodb-index.test,
	innodb.result, innodb.test:
	Disallow a duplicate index name when creating an index.

2010-02-11	The InnoDB Team

	* include/mem0mem.h, include/mem0mem.ic, mem/mem0mem.c:
	Fix Bug#49535 Available memory check slows down crash
	recovery tens of times

2010-02-09	The InnoDB Team

	* buf/buf0buf.c:
	Fix Bug#38901 InnoDB logs error repeatedly when trying to load
	page into buffer pool

2010-02-09	The InnoDB Team

	* srv/srv0srv.c:
	Let the master thread sleep if the amount of work to be done is
	calibrated as taking less than a second.

2010-02-04	The InnoDB Team

	* btr/btr0btr.c, btr/btr0cur.c, btr/btr0pcur.c, buf/buf0buf.c,
	include/btr0btr.h, include/btr0cur.h, include/btr0pcur.h,
	include/btr0pcur.ic, include/buf0buf.h, row/row0ins.c, row/row0sel.c:
	Pass the file name and line number of the caller of the
	b-tree cursor functions to the buffer pool requests, in order
	to make the latch diagnostics more accurate.

2010-02-03	The InnoDB Team

	* lock/lock0lock.c:
	Fix Bug#49001 SHOW INNODB STATUS deadlock info incorrect
	when deadlock detection aborts

2010-02-03	The InnoDB Team

	* buf/buf0lru.c:
	Fix Bug#35077 Very slow DROP TABLE (ALTER TABLE, OPTIMIZE TABLE)
	on compressed tables

2010-02-03	The InnoDB Team

	* handler/ha_innodb.cc, include/row0mysql.h, row/row0mysql.c:
	Clean up CHECK TABLE error handling.

2010-02-01	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb-autoinc.test,
	mysql-test/innodb-autoinc.result,
	mysql-test/innodb-autoinc-44030.test,
	mysql-test/innodb-autoinc-44030.result:
	Fix Bug#49497 Error 1467 (ER_AUTOINC_READ_FAILED) on inserting
	a negative value

2010-01-27	The InnoDB Team

	* include/row0mysql.h, log/log0recv.c, row/row0mysql.c:
	Drop temporary tables at startup.
	This addresses the third aspect of
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.

2010-01-21	The InnoDB Team

	* buf/buf0buf.c:
	Do not merge buffered inserts to compressed pages before
	the redo log has been applied in crash recovery.

2010-01-13	The InnoDB Team

	* row/row0sel.c:
	On the READ UNCOMMITTED isolation level, do not attempt to access
	a clustered index record that has been marked for deletion. The
	built-in InnoDB in MySQL 5.1 and earlier would attempt to retrieve
	a previous version of the record in this case.

2010-01-13	The InnoDB Team

	* buf/buf0buf.c:
	When disabling the adaptive hash index, check the block state
	before checking block->is_hashed, because the latter may be
	uninitialized right after server startup.

2010-01-12	The InnoDB Team

	* handler/ha_innodb.cc, handler/ha_innodb.h:
	Fix Bug#46193 crash when accessing tables after enabling
	innodb_force_recovery option

2010-01-12	The InnoDB Team

	* row/row0mysql.c:
	Fix Bug#49238 Creating/Dropping a temporary table while at 1023
	transactions will cause assert.

2009-12-02	The InnoDB Team

	* srv/srv0start.c:
	Display the zlib version number at startup.
	InnoDB compressed tables use zlib, and the implementation depends
	on the zlib function compressBound(), whose definition was slightly
	changed in zlib version 1.2.3.1 in 2006. MySQL bundles zlib 1.2.3
	from 2005, but some installations use a more recent zlib.

2009-11-30	The InnoDB Team

	* dict/dict0crea.c, dict/dict0mem.c, dict/dict0load.c,
	dict/dict0boot.c, fil/fil0fil.c, handler/ha_innodb.cc,
	include/dict0mem.h, row/row0mysql.c:
	Fix the bogus warning messages for non-existing temporary
	tables that were reported in
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.
	The actual crash recovery bug was corrected on 2009-04-29.

2009-11-27	The InnoDB Team

	InnoDB Plugin 1.0.6 released

2009-11-20	The InnoDB Team

	* handler/ha_innodb.cc:
	Add a workaround to prevent a crash due to Bug#45961 DDL on
	partitioned innodb tables leaves data dictionary in an inconsistent
	state

2009-11-19	The InnoDB Team

	* btr/btr0btr.c:
	Fix Bug#48469 when innodb tablespace is configured too small, crash
	and corruption!

2009-11-19	The InnoDB Team

	* data/data0type.c:
	Fix Bug#48526 Data type for float and double is incorrectly reported
	in InnoDB table monitor


mysql-test/include/ibd2sdi.pl
use strict;
use warnings;
use File::Copy;
use MIME::Base64 qw(decode_base64);

my $json_file_path = $ENV{'JSON_FILE_PATH'} or die;
my $dir = $ENV{'MYSQLTEST_VARDIR'} or die;


sub ibd2sdi_replace() {
  # open file in write mode
  open IN_FILE,"<", "$json_file_path" or die $! ;
  open (OUT_FILE, ">:raw", "$dir/tmp/tmpfile") or die $! ;
  while(<IN_FILE>) {

    # Remove number output from "last_altered" value
    $_=~ s/("last_altered": )[0-9]+/$1NNN/g;

    # Remove number output from "created" value
    $_=~ s/("created": )[0-9]+/$1NNN/g;

    # Remove se_private_data: id & trx_id output. retain the root page number value
    $_=~ s/("se_private_data":) "id=[0-9]+;root=[0-9]+;space_id=[0-9]+;table_id=[0-9]+;trx_id=[0-9]+;"/$1 "id=A;root=B;space_id=C;table_id=D;trx_id=E"/g;

    # Remove se_private_data: id & trx_id output. retain the root page number value
    $_=~ s/("se_private_data":) "id=[0-9]+;root=[0-9]+;trx_id=[0-9]+;"/$1 "id=X;root=Y;trx_id=Z"/g;

    # Remove se_private_data: table_id.
    $_=~ s/("se_private_data":) "table_id=[0-9]+;"/$1 "table_id=X"/g;

    # Remove se_private_data: physical_pos & table_id.
    $_=~ s/("se_private_data":) "physical_pos=[0-9]+;table_id=[0-9]+;"/$1 "physical_pos=X;table_id=Y"/g;

    # Remove se_private_data: id.
    $_=~ s/("se_private_data":) "id=[0-9]+;"/$1 "id=X"/g;

    # Remove se_private_id: id. This is innodb table_id
    $_=~ s/("se_private_id":) [0-9]+/$1NNN/g;

    # Remove object_id output
    $_=~ s/("id": )[0-9]+/$1X/g;

    # Remove file_per_tablespace id output in dd::Table
    $_=~ s/("tablespace_ref": "innodb_file_per_table.)[0-9]+"/$1X"/g;

    # Remove file_per_tablespace id output in dd::Tablespace
    $_=~ s/("name": "innodb_file_per_table.)[0-9]+"/$1X"/g;

    # Remove id output in se_prvate_data of dd::Tablespace
    $_=~ s/("se_private_data": "flags=)([0-9]+)(;id=)([0-9]+)(;server_version=)([0-9]+)(;space_version=)([0-9]+);/$1X$3Y$5Z$7M;/g;

    # Remove any version id
    $_=~ s/("[a-z_]+_version_id": )[0-9]+/$1X/g;

    # Remove any version
    $_=~ s/("[a-z_]+_version": )[0-9]+/$1X/g;

    # Remove se_private_data: null and table_id
    $_=~ s/("se_private_data":) "default_null=[0-1];table_id=[0-9]+;"/$1 "default_null=X;table_id=Y"/g;

    # Remove extra path separator seen on windows and partition names
    if (m/"filename":/)
    {
      $_=~ s/\\\\/\//g;
    }

    $_=~ s/#P#/#p#/g;

    $_=~ s/#SP#/#sp#/g;

    $_=~ s/("default_value": ).*/$1"",/g;

    # Names of SET elements in the input JSON are Base64-encoded.
    # For the easier reading, append a fake "decoded-name": "foo" name/value pair
    # to the output JSON, where "foo" is for a decoded SET element name.
    $_=~ s/^ {24}"name": "([^"]+)",$/$& . ' "decoded-name": "' . decode_base64($1) . '",'/e;

    print OUT_FILE $_;
  }
  close(IN_FILE);
  close(OUT_FILE);
  #move the new content from tmp file to the orginal file.
  move ("$dir/tmp/tmpfile", "$json_file_path");
}

sub ibd2sdi_replace_system() {
  # open file in write mode
  open IN_FILE,"<", "$json_file_path" or die $! ;
  open (OUT_FILE, ">:raw", "$dir/tmp/tmpfile") or die $! ;
  while(<IN_FILE>) {

    # Remove number output from "last_altered" value
    $_=~ s/("last_altered": )[0-9]+/$1NNN/g;

    # Remove number output from "created" value
    $_=~ s/("created": )[0-9]+/$1NNN/g;

    # Remove se_private_data: id & trx_id output. retain the root page number value
    # This is only thing that differed from normal replace
    $_=~ s/("se_private_data":) "id=[0-9]+;root=[0-9]+;space_id=[0-9]+;table_id=[0-9]+;trx_id=[0-9]+;"/$1 "id=A;root=B;space_id=C;table_id=D;trx_id=E"/g;

    # Remove se_private_data: id & trx_id output. retain the root page number value
    $_=~ s/("se_private_data":) "id=[0-9]+;root=[0-9]+;trx_id=[0-9]+;"/$1 "id=X;root=Y;trx_id=Z"/g;

    # Remove se_private_data: table_id.
    $_=~ s/("se_private_data":) "table_id=[0-9]+;"/$1 "table_id=X"/g;

    # Remove se_private_data: physical_pos & table_id.
    $_=~ s/("se_private_data":) "physical_pos=[0-9]+;table_id=[0-9]+;"/$1 "physical_pos=X;table_id=Y"/g;

    # Remove se_private_data: id.
    $_=~ s/("se_private_data":) "id=[0-9]+;"/$1 "id=X"/g;

    # Remove se_private_id: id. This is innodb table_id
    $_=~ s/("se_private_id":) [0-9]+/$1NNN/g;

    # Remove object_id output
    $_=~ s/("id": )[0-9]+/$1X/g;

    # Remove file_per_tablespace id output in dd::Table
    $_=~ s/("tablespace_ref": "innodb_file_per_table.)[0-9]+"/$1X"/g;

    # Remove file_per_tablespace id output in dd::Tablespace
    $_=~ s/("name": "innodb_file_per_table.)[0-9]+"/$1X"/g;

    # Remove id output in se_prvate_data of dd::Tablespace
    $_=~ s/("se_private_data": "flags=)([0-9]+)(;id=)([0-9]+)(;server_version=)([0-9]+)(;space_version=)([0-9]+);/$1X$3Y$5Z$7M;/g;

    # Remove any version id
    $_=~ s/("[a-z_]+_version_id": )[0-9]+/$1X/g;


    # Remove any version
    $_=~ s/("[a-z_]+_version": )[0-9]+/$1X/g;

    # Remove extra path separator seen on windows and partition names
    if (m/"filename":/)
    {
      $_=~ s/\\\\/\//g;
    }

    $_=~ s/#P#/#p#/g;

    $_=~ s/#SP#/#sp#/g;

    $_=~ s/("default_value": ).*/$1"",/g;

    # Names of SET elements in the input JSON are Base64-encoded.
    # For the easier reading, append a fake "decoded-name": "foo" name/value pair
    # to the output JSON, where "foo" is for a decoded SET element name.
    $_=~ s/^ {24}"name": "([^"]+)",$/$& . ' "decoded-name": "' . decode_base64($1) . '",'/e;

    print OUT_FILE $_;
  }
  close(IN_FILE);
  close(OUT_FILE);
  #move the new content from tmp file to the orginal file.
  move ("$dir/tmp/tmpfile", "$json_file_path");
}

sub ibd2sdi_replace_mysql() {
  # open file in write mode
  open IN_FILE,"<", "$json_file_path" or die $! ;
  open (OUT_FILE, ">:raw", "$dir/tmp/tmpfile") or die $! ;
  while(<IN_FILE>) {

    # Remove number output from "last_altered" value
    $_=~ s/("last_altered": )[0-9]+/$1NNN/g;

    # Remove number output from "created" value
    $_=~ s/("created": )[0-9]+/$1NNN/g;

    # Remove se_private_data: id & trx_id output. retain the root page number value
    $_=~ s/("se_private_data":) "id=[0-9]+;root=[0-9]+;space_id=[0-9]+;table_id=[0-9]+;trx_id=[0-9]+;"/$1 "id=A;root=B;space_id=C;table_id=D;trx_id=E"/g;

    # Remove se_private_data: id & trx_id output. retain the root page number value
    $_=~ s/("se_private_data":) "id=[0-9]+;root=[0-9]+;trx_id=[0-9]+;"/$1 "id=A;root=B;trx_id=C"/g;

    # Remove se_private_data: table_id.
    $_=~ s/("se_private_data":) "table_id=[0-9]+;"/$1 "table_id=X"/g;

    # Remove se_private_data: physical_pos & table_id.
    $_=~ s/("se_private_data":) "physical_pos=[0-9]+;table_id=[0-9]+;"/$1 "physical_pos=X;table_id=Y"/g;

    # Remove se_private_data: id.
    $_=~ s/("se_private_data":) "id=[0-9]+;"/$1 "id=X"/g;

    # Remove se_private_id: id. This is innodb table_id
    $_=~ s/("se_private_id":) [0-9]+/$1NNN/g;

    # Remove object_id output
    $_=~ s/("id": )[0-9]+/$1X/g;

    # Remove file_per_tablespace id output in dd::Table
    $_=~ s/("tablespace_ref": "innodb_file_per_table.)[0-9]+"/$1X"/g;

    # Remove file_per_tablespace id output in dd::Tablespace
    $_=~ s/("name": "innodb_file_per_table.)[0-9]+"/$1X"/g;

    # Remove id output in se_prvate_data of dd::Tablespace
    $_=~ s/("se_private_data": "flags=)([0-9]+)(;id=)([0-9]+)(;server_version=)([0-9]+)(;space_version=)([0-9]+);/$1X$3Y$5Z$7M;/g;

    # Remove any version id
    $_=~ s/("[a-z_]+_version_id": )[0-9]+/$1X/g;

    # Remove any version
    $_=~ s/("[a-z_]+_version": )[0-9]+/$1X/g;

    # Remove extra path separator seen on windows and partition names
    if (m/"filename":/)
    {
      $_=~ s/\\\\/\//g;
    }

    $_=~ s/#P#/#p#/g;

    $_=~ s/#SP#/#sp#/g;

    $_=~ s/("default_value": ).*/$1"",/g;

    $_=~ s/("collation_id": )[0-9]+/$1X/g;

    # Names of SET elements in the input JSON are Base64-encoded.
    # For the easier reading, append a fake "decoded-name": "foo" name/value pair
    # to the output JSON, where "foo" is for a decoded SET element name.
    $_=~ s/^ {24}"name": "([^"]+)",$/$& . ' "decoded-name": "' . decode_base64($1) . '",'/e;


    print OUT_FILE $_;
  }

  close(IN_FILE);
  close(OUT_FILE);
  #move the new content from tmp file to the orginal file.
  move ("$dir/tmp/tmpfile", "$json_file_path");
}

# "require ibd2sdi.pl" says the module should return true and this
# achieved by the below statement
1;
